File: mobile/modules/bluetooth-sdk/ios/Source/tts/TTSTools.swift
import Foundation

/// Utilities for offline Sherpa-ONNX Supertonic 3 TTS.
///
/// Supertonic 3 is a single multilingual model bundle (31 languages) with
/// 10 preset voice styles packed into voice.bin. Speaker IDs are indexed
/// alphabetically by JSON filename, so the order is F1, F2, F3, F4, F5,
/// M1, M2, M3, M4, M5 (sids 0..9). We default to F1 (sid 0) for a female
/// voice.
class TTSTools {
    private static let kFileDurationPredictor = "duration_predictor.int8.onnx"
    private static let kFileTextEncoder = "text_encoder.int8.onnx"
    private static let kFileVectorEstimator = "vector_estimator.int8.onnx"
    private static let kFileVocoder = "vocoder.int8.onnx"
    private static let kFileTtsJson = "tts.json"
    private static let kFileUnicodeIndexer = "unicode_indexer.bin"
    private static let kFileVoiceStyle = "voice.bin"

    private static let kRequiredFiles: [String] = [
        kFileDurationPredictor,
        kFileTextEncoder,
        kFileVectorEstimator,
        kFileVocoder,
        kFileTtsJson,
        kFileUnicodeIndexer,
        kFileVoiceStyle,
    ]

    private static let kSupportedLangs: Set<String> = [
        "en", "ko", "ja", "ar", "bg", "cs", "da", "de", "el", "es", "et",
        "fi", "fr", "hi", "hr", "hu", "id", "it", "lt", "lv", "nl", "pl",
        "pt", "ro", "ru", "sk", "sl", "sv", "tr", "uk", "vi",
    ]

    static func setTtsModelDetails(_ path: String, _ languageCode: String) {
        UserDefaults.standard.set(path, forKey: "TTSModelPath")
        UserDefaults.standard.set(languageCode, forKey: "TTSModelLanguageCode")
        UserDefaults.standard.synchronize()
    }

    static func getTtsModelPath() -> String {
        return UserDefaults.standard.string(forKey: "TTSModelPath") ?? ""
    }

    static func getTtsModelLanguage() -> String {
        return UserDefaults.standard.string(forKey: "TTSModelLanguageCode") ?? "en-US"
    }

    static func checkTTSModelAvailable() -> Bool {
        guard let modelPath = UserDefaults.standard.string(forKey: "TTSModelPath") else {
            return false
        }
        return validateTTSModel(modelPath)
    }

    static func validateTTSModel(_ path: String) -> Bool {
        let fileManager = FileManager.default
        var isDirectory: ObjCBool = false
        if !fileManager.fileExists(atPath: path, isDirectory: &isDirectory) || !isDirectory.boolValue {
            Bridge.log("TTS model path does not exist or is not a directory: \(path)")
            return false
        }

        for name in kRequiredFiles {
            let filePath = (path as NSString).appendingPathComponent(name)
            if !fileManager.fileExists(atPath: filePath) {
                Bridge.log("TTS model missing required file '\(name)' at: \(path)")
                return false
            }
            let size = (try? fileManager.attributesOfItem(atPath: filePath)[.size] as? UInt64) ?? 0
            if size == 0 {
                Bridge.log("TTS model file '\(name)' is empty at: \(path)")
                return false
            }
        }

        return true
    }

    static func generateTtsAudio(
        text: String,
        modelPath: String,
        outputPath: String,
        speakerId: Int,
        speed: Double
    ) -> Bool {
        guard !text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            Bridge.log("TTS_ERROR: text is empty")
            return false
        }
        guard validateTTSModel(modelPath) else {
            Bridge.log("TTS_ERROR: model is invalid: \(modelPath)")
            return false
        }

        do {
            let outputURL = URL(fileURLWithPath: outputPath)
            try FileManager.default.createDirectory(
                at: outputURL.deletingLastPathComponent(),
                withIntermediateDirectories: true,
                attributes: nil
            )

            let supertonic = sherpaOnnxOfflineTtsSupertonicModelConfig(
                durationPredictor: (modelPath as NSString).appendingPathComponent(kFileDurationPredictor),
                textEncoder: (modelPath as NSString).appendingPathComponent(kFileTextEncoder),
                vectorEstimator: (modelPath as NSString).appendingPathComponent(kFileVectorEstimator),
                vocoder: (modelPath as NSString).appendingPathComponent(kFileVocoder),
                ttsJson: (modelPath as NSString).appendingPathComponent(kFileTtsJson),
                unicodeIndexer: (modelPath as NSString).appendingPathComponent(kFileUnicodeIndexer),
                voiceStyle: (modelPath as NSString).appendingPathComponent(kFileVoiceStyle)
            )
            var modelConfig = sherpaOnnxOfflineTtsModelConfig(
                numThreads: 8,
                supertonic: supertonic
            )
            var config = sherpaOnnxOfflineTtsConfig(model: modelConfig, maxNumSentences: 1)
            let tts = SherpaOnnxOfflineTtsWrapper(config: &config)

            // The new wrapper dropped numSpeakers() in v1.13.2; call the C API
            // directly. The C signature accepts the underlying OpaquePointer.
            let numSpeakers = max(1, Int(SherpaOnnxOfflineTtsNumSpeakers(tts.tts)))
            let sid = max(0, min(speakerId, numSpeakers - 1))
            let clampedSpeed = Float(min(max(speed, 0.5), 2.0))
            let lang = languageTagToSupertonicLang(getTtsModelLanguage())

            // generate(...) no longer accepts `extra`; that field moved onto
            // SherpaOnnxGenerationConfigSwift and is delivered via
            // generateWithConfig(...).
            //
            // numSteps controls the Supertonic flow-matching vocoder's
            // iteration count. The Swift wrapper defaults to 1, which produces
            // unintelligible robot audio — sherpa-onnx's actual C-side default
            // is 5. Pin it here so we don't inherit the wrapper's bug.
            let genConfig = SherpaOnnxGenerationConfigSwift(
                speed: clampedSpeed,
                sid: sid,
                numSteps: 5,
                extra: ["lang": lang]
            )
            let audio = tts.generateWithConfig(
                text: text,
                config: genConfig,
                callback: nil,
                arg: nil
            )
            let saved = audio.save(filename: outputPath) == 1
            Bridge.log(
                "TTS generated \(outputPath): saved=\(saved) sid=\(sid) lang=\(lang) " +
                    "sr=\(audio.sampleRate) samples=\(audio.n) speed=\(clampedSpeed)"
            )
            return saved
        } catch {
            Bridge.log("TTS_ERROR: \(error.localizedDescription)")
            return false
        }
    }

    /// Convert a BCP-47 tag (e.g. "en-US", "fr-FR") to a Supertonic 2-letter code.
    private static func languageTagToSupertonicLang(_ tag: String) -> String {
        let primary = tag.split(separator: "-").first.map(String.init)?.lowercased() ?? "en"
        return kSupportedLangs.contains(primary) ? primary : "en"
    }
}

File: mobile/modules/bluetooth-sdk/android/src/main/java/com/mentra/bluetoothsdk/tts/TTSTools.kt
package com.mentra.core.tts

import android.content.Context
import android.content.SharedPreferences
import com.k2fsa.sherpa.onnx.GenerationConfig
import com.k2fsa.sherpa.onnx.OfflineTts
import com.k2fsa.sherpa.onnx.OfflineTtsConfig
import com.k2fsa.sherpa.onnx.OfflineTtsModelConfig
import com.k2fsa.sherpa.onnx.OfflineTtsSupertonicModelConfig
import com.mentra.bluetoothsdk.Bridge
import java.io.File
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledFuture
import java.util.concurrent.TimeUnit

/**
 * Utilities for offline Sherpa-ONNX Supertonic 3 TTS model management.
 *
 * Supertonic 3 is a single multilingual model bundle (31 languages) with 10
 * preset voice styles packed into a single voice.bin. Speaker IDs are indexed
 * alphabetically by JSON filename, so the order is F1, F2, F3, F4, F5, M1, M2,
 * M3, M4, M5 (sids 0..9). We default to F1 (sid 0) for a female voice.
 */
object TTSTools {
    private const val PREFS_NAME = "MentraPrefs"
    private const val KEY_TTS_MODEL_PATH = "TTSModelPath"
    private const val KEY_TTS_MODEL_LANGUAGE = "TTSModelLanguageCode"

    private const val FILE_DURATION_PREDICTOR = "duration_predictor.int8.onnx"
    private const val FILE_TEXT_ENCODER = "text_encoder.int8.onnx"
    private const val FILE_VECTOR_ESTIMATOR = "vector_estimator.int8.onnx"
    private const val FILE_VOCODER = "vocoder.int8.onnx"
    private const val FILE_TTS_JSON = "tts.json"
    private const val FILE_UNICODE_INDEXER = "unicode_indexer.bin"
    private const val FILE_VOICE_STYLE = "voice.bin"

    /**
     * Keep the native model warm briefly after use to reduce repeated TTS latency while
     * still reclaiming its memory when speech has been idle for a while.
     */
    private const val MODEL_IDLE_TIMEOUT_MINUTES = 3L
    private val modelLock = Any()
    private val teardownExecutor =
            Executors.newSingleThreadScheduledExecutor { runnable ->
                Thread(runnable, "MentraTtsIdleTeardown").apply { isDaemon = true }
            }
    private var cachedTts: OfflineTts? = null
    private var cachedModelPath: String? = null
    private var teardownFuture: ScheduledFuture<*>? = null
    private var teardownGeneration = 0L

    private val REQUIRED_FILES = listOf(
            FILE_DURATION_PREDICTOR,
            FILE_TEXT_ENCODER,
            FILE_VECTOR_ESTIMATOR,
            FILE_VOCODER,
            FILE_TTS_JSON,
            FILE_UNICODE_INDEXER,
            FILE_VOICE_STYLE,
    )

    fun setTtsModelDetails(context: Context, path: String, languageCode: String) {
        synchronized(modelLock) {
            if (cachedModelPath != null && cachedModelPath != path) {
                releaseCachedModelLocked("model path changed")
            }
        }
        val prefs = getPrefs(context)
        prefs.edit().apply {
            putString(KEY_TTS_MODEL_PATH, path)
            putString(KEY_TTS_MODEL_LANGUAGE, languageCode)
            apply()
        }
        Bridge.log("TTS model details saved: path=$path, language=$languageCode")
    }

    fun getTtsModelPath(context: Context): String {
        val prefs = getPrefs(context)
        return prefs.getString(KEY_TTS_MODEL_PATH, "") ?: ""
    }

    fun getTtsModelLanguage(context: Context): String {
        val prefs = getPrefs(context)
        return prefs.getString(KEY_TTS_MODEL_LANGUAGE, "en-US") ?: "en-US"
    }

    fun checkTTSModelAvailable(context: Context): Boolean {
        val modelPath = getTtsModelPath(context)
        if (modelPath.isEmpty()) {
            return false
        }
        return validateTTSModel(modelPath)
    }

    fun validateTTSModel(path: String): Boolean {
        val modelDir = File(path)
        if (!modelDir.exists() || !modelDir.isDirectory) {
            Bridge.log("TTS model path does not exist or is not a directory: $path")
            return false
        }

        for (name in REQUIRED_FILES) {
            val file = File(modelDir, name)
            if (!file.exists() || !file.canRead() || file.length() == 0L) {
                Bridge.log("TTS model missing required file '$name' at: $path")
                return false
            }
        }

        return true
    }

    private val SUPPORTED_LANGS = setOf(
            "en", "ko", "ja", "ar", "bg", "cs", "da", "de", "el", "es", "et",
            "fi", "fr", "hi", "hr", "hu", "id", "it", "lt", "lv", "nl", "pl",
            "pt", "ro", "ru", "sk", "sl", "sv", "tr", "uk", "vi",
    )

    /** Convert a BCP-47 tag like "en-US" or "fr-FR" to a Supertonic 2-letter code. */
    private fun languageTagToSupertonicLang(tag: String?): String {
        if (tag.isNullOrBlank()) return "en"
        val primary = tag.substringBefore('-').lowercase()
        return if (primary in SUPPORTED_LANGS) primary else "en"
    }

    fun generateTtsAudio(
            context: Context,
            text: String,
            modelPath: String,
            outputPath: String,
            speakerId: Int,
            speed: Float,
    ): Boolean {
        val languageTag = getTtsModelLanguage(context)
        if (text.isBlank()) {
            Bridge.log("TTS_ERROR: text is empty")
            return false
        }
        if (!validateTTSModel(modelPath)) {
            Bridge.log("TTS_ERROR: model is invalid: $modelPath")
            return false
        }

        val outputFile = File(outputPath)
        outputFile.parentFile?.mkdirs()

        return synchronized(modelLock) {
            try {
                val tts = getOrCreateModelLocked(modelPath)

                // sid out of range falls back to F1 (sid 0); clamp 0..9.
                val sid = speakerId.coerceIn(0, tts.numSpeakers() - 1)
                val clampedSpeed = speed.coerceIn(0.5f, 2.0f)
                val lang = languageTagToSupertonicLang(languageTag)
                val genConfig = GenerationConfig(
                        sid = sid,
                        speed = clampedSpeed,
                        numSteps = 5,
                        extra = mapOf("lang" to lang),
                )
                val audio = tts.generateWithConfig(text = text, config = genConfig)
                val saved = audio.save(outputFile.absolutePath)
                Bridge.log("TTS generated ${outputFile.absolutePath}: saved=$saved sid=$sid lang=$lang")
                saved
            } catch (e: Exception) {
                Bridge.log("TTS_ERROR: ${e.javaClass.simpleName}: ${e.message}")
                e.printStackTrace()
                false
            } finally {
                scheduleIdleTeardownLocked()
            }
        }
    }

    private fun getOrCreateModelLocked(modelPath: String): OfflineTts {
        cachedTts?.let { cached ->
            if (cachedModelPath == modelPath) {
                Bridge.log("TTS reusing cached model")
                return cached
            }
            releaseCachedModelLocked("model path changed")
        }

        val modelDir = File(modelPath)
        val supertonic = OfflineTtsSupertonicModelConfig(
                durationPredictor = File(modelDir, FILE_DURATION_PREDICTOR).absolutePath,
                textEncoder = File(modelDir, FILE_TEXT_ENCODER).absolutePath,
                vectorEstimator = File(modelDir, FILE_VECTOR_ESTIMATOR).absolutePath,
                vocoder = File(modelDir, FILE_VOCODER).absolutePath,
                ttsJson = File(modelDir, FILE_TTS_JSON).absolutePath,
                unicodeIndexer = File(modelDir, FILE_UNICODE_INDEXER).absolutePath,
                voiceStyle = File(modelDir, FILE_VOICE_STYLE).absolutePath,
        )
        val modelConfig = OfflineTtsModelConfig(
                supertonic = supertonic,
                numThreads = 8,
                provider = "cpu",
        )
        val config = OfflineTtsConfig(
                model = modelConfig,
                maxNumSentences = 1,
                silenceScale = 0.2f,
        )
        return OfflineTts(null, config).also {
            cachedTts = it
            cachedModelPath = modelPath
            Bridge.log("TTS loaded model; keeping it warm for $MODEL_IDLE_TIMEOUT_MINUTES minutes after use")
        }
    }

    private fun scheduleIdleTeardownLocked() {
        if (cachedTts == null) return
        teardownFuture?.cancel(false)
        val generation = ++teardownGeneration
        teardownFuture =
                teardownExecutor.schedule(
                        {
                            synchronized(modelLock) {
                                if (generation == teardownGeneration) {
                                    releaseCachedModelLocked("idle for $MODEL_IDLE_TIMEOUT_MINUTES minutes")
                                }
                            }
                        },
                        MODEL_IDLE_TIMEOUT_MINUTES,
                        TimeUnit.MINUTES,
                )
    }

    private fun releaseCachedModelLocked(reason: String) {
        teardownFuture?.cancel(false)
        teardownFuture = null
        teardownGeneration++
        val tts = cachedTts
        cachedTts = null
        cachedModelPath = null
        if (tts == null) return
        try {
            tts.release()
            Bridge.log("TTS released cached model: $reason")
        } catch (e: Exception) {
            Bridge.log("TTS release failed: ${e.message}")
        }
    }

    private fun getPrefs(context: Context): SharedPreferences {
        return context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    }
}
